kubespawner/spawner.py

```diff
@@ -3011,69 +3011,97 @@ def _options_from_form(self, formdata):
             user_options (dict): the selected profile in the user_options form,
                 e.g. ``{"profile": "cpus-8"}``
         """
-        profile = formdata.get('profile', [None])[0]
+        profile_slug = formdata.get('profile', [None])[0]
 
-        options = {'profile': profile}
+        # initialize a dictionary to return
+        user_options = {}
 
-        # Load any options if they are set for this profile, and this profile only
-        # In the default template we use, all form options for a particular profile
-        # come through of the format 'profile-option-{profile}-{option-slug}'
+        # if a profile is declared, add a dictionary key for the profile, and
+        # dictionary keys for the formdata related to the profile's
+        # profile_options, as recognized by being named like:
         #
-        # FIXME: Since we only return recognised formdata in user_options, we
-        #        shouldn't need later validation of this, and we need to emit
-        #        warnings here rather than later about unrecognised options.
+        #     profile-option-{profile_slug}--{profile_option_slug}
         #
-        if profile:
-            option_formdata_prefix = f'profile-option-{profile}--'
+        if profile_slug:
+            user_options["profile"] = profile_slug
+            prefix = f'profile-option-{profile_slug}--'
             for k, v in formdata.items():
-                if k.startswith(option_formdata_prefix):
-                    stripped_key = k[len(option_formdata_prefix) :]
-                    options[stripped_key] = v[0]
+                if k.startswith(prefix):
+                    profile_option_slug = k[len(prefix) :]
+                    user_options[profile_option_slug] = v[0]
+
+        # warn about any unrecognized form data, which is anything besides
+        # "profile" and "profile-option-" prefixed keys
+        unrecognized_keys = set(formdata)
+        unrecognized_keys = unrecognized_keys.difference({"profile"})
+        unrecognized_keys = [
+            k for k in unrecognized_keys if not k.startswith("profile-option-")
+        ]
+        if unrecognized_keys:
+            self.log.warning(
+                "Ignoring unrecognized form data in spawn request: %s",
+                ", ".join(map(str, sorted(unrecognized_keys))),
+            )
```

Comment:
This logic existed elsewhere before in a broken state. It was broken before becaue it looked at parsed user_options, but had logic about the formdata that is to be parsed to user_options.